src/cpp/core/FileSerializer.cpp

```diff
@@ -24,13 +24,63 @@
 #include <boost/algorithm/string/trim.hpp>
 #include <boost/algorithm/string/predicate.hpp>
 #include <boost/iostreams/copy.hpp>
+#include <boost/thread.hpp>
 
 #include <shared_core/FilePath.hpp>
+#include <core/DateTime.hpp>
 #include <core/StringUtils.hpp>
 
 namespace rstudio {
 namespace core {
 
+namespace {
+
+bool isFileLockedError(const Error& error)
+{
+   // exclusive file access is only present on Windows
+#ifndef _WIN32
+   return false;
+#endif
+   return (error && error.getCode() == ERROR_SHARING_VIOLATION);
+}
+
+Error openFileForWritingWithRetry(const FilePath& filePath,
+                                  bool truncate,
+                                  int maxOpenRetrySeconds,
+                                  std::shared_ptr<std::ostream>* pOfs)
+{
+   using namespace boost::posix_time;
+
+   bool opened = false;
+   ptime startTime = second_clock::universal_time();
+   Error lastError;
+
+   // do not allow negative values - regular signed int was chosen here for
+   // easier integration with other parts of the codebase
+   if (maxOpenRetrySeconds < 0)
+      maxOpenRetrySeconds = 0;
+
+   while (!opened)
```

Comment:
It looks like this loop doesn't update `opened` below?

Reply:
Thanks for that - this was some leftovers from my prior impl that was using `do{}while();`. Changed this to just be a `while(true)` since it either exits the loop via return or a break when the time has elapsed.